### PR TITLE
More trace logging for Dataset Service

### PR DESCRIPTION
**Pull request by anew**
- log all requests in dataset service client and handler
- add a header with the callers thread name and log that on the server side 
- log requests in netty pipeline before they are routed to a handler
- log thread dump in dataset service client in case of timeout
- misc minor fixes

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java [hunk 1]
```diff
@@ -367,4 +375,83 @@ private HttpResponse doRequest(HttpRequest.Builder requestBuilder) throws Datase
     }
     return builder.addHeader(Constants.Security.Headers.USER_ID, userId);
   }
+
+  private static void logThreadDump() {
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Timeout while making request to dataset service. Thread dump follows:");
+      ThreadMXBean bean = ManagementFactory.getThreadMXBean();
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: How is this comparing to `Thread.getAllStackTraces`?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java [hunk 2]
```diff
@@ -178,17 +178,16 @@ public void setMetadataPublisher(MetadataPublisher metadataPublisher) {
...
    * @param instance instance to get
-   * @param owners the {@link EntityId entities} that will be using the dataset instance
    * @return the dataset instance's {@link DatasetMeta}
    * @throws NotFoundException if either the namespace or dataset instance is not found,
    * @throws IOException if there is a problem in making an HTTP request to check if the namespace exists
    * @throws UnauthorizedException if perimeter security and authorization are enabled, and the current user does not
    *  have any privileges on the #instance
    */
-  DatasetMeta get(final DatasetId instance, List<? extends EntityId> owners) throws Exception {
+  DatasetMeta get(final DatasetId instance,  List<? extends EntityId> owners) throws Exception {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: unnecessary change?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetService.java [hunk 3]
```diff
@@ -73,12 +78,29 @@ public DatasetService(CConfiguration cConf,
...
+                        DatasetInstanceService datasetInstanceService) {
     this.cConf = cConf;
     this.typeService = datasetTypeService;
     DatasetTypeHandler datasetTypeHandler = new DatasetTypeHandler(datasetTypeService);
     DatasetInstanceHandler datasetInstanceHandler = new DatasetInstanceHandler(datasetInstanceService);
     NettyHttpService.Builder builder = new CommonNettyHttpServiceBuilder(cConf, Constants.Service.DATASET_MANAGER)
+      .addChannelPipelineModifier(new ChannelPipelineModifier() {
+        @Override
+        public void modify(ChannelPipeline channelPipeline) {
+          if (LOG.isTraceEnabled()) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetService.java]: Might be better not to add the modifier if log trace is not enabled.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java (hunk 1 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: not sure I copied this from a branch that @albertshau had created earlier. But I believe a stack trace only contains the class and line number of each element; whereas this gives us more information such as the state of the thread, what locks it is holding/waiting on, etc. That would help us identify deadlocks.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java (hunk 2 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: unintentional, too :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetService.java (hunk 3 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetService.java]: I was thinking that with Twill, we change the log level after a container has started. Do you think it is a performance concern?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java [hunk 4]
```diff
@@ -367,4 +375,83 @@ private HttpResponse doRequest(HttpRequest.Builder requestBuilder) throws Datase
...
+      for (ThreadInfo threadInfo : bean.dumpAllThreads(true, true)) {
+        LOG.trace("{}", toString(threadInfo));
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: can we have all of the threadinfos in one log message, instead of split across multiple log messages?
This will make it easier to correlate them.
For instance, there may be two threads that are both calling the `logThreadDump` method, and their log messages get intertwined.

**Comment by anew** [comment]
@chtyim @anwar6953 I have addressed your comments. Please take another look

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java (hunk 4 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: good point, done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java (hunk 1 above)
Comment (chtyim) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java]: I think the getAllStackTraces return a Map from Thread to StackElement[], which is kind of similar to this. I was just curious :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetService.java (hunk 3 above)
Comment (chtyim) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetService.java]: That's all right, shouldn't be too big of a concern.

**Review by chtyim — APPROVED** [review]
LGTM
